aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1602,6 +1602,69 @@ Tensor& linalg_norm_out(Tensor& result, const Tensor& self, std::string ord, opt
 }
 
+Tensor linalg_tensorsolve(const Tensor& self, const Tensor& other, optional<IntArrayRef> dims) {
+  /*
+  The idea is to reduce the problem to 2D matrix solve.
+  Step 1. (optional) `self` is permuted with `dims` such that dimensions from `dims` are moved to the right.
+  For example, if we have 4D input with the shape (1, 2, 3, 4) and dims=(0, 2),
+  then the result of permutation would have the shape (2, 4, 1, 3).
+  Step 2. reshape `self` to 2D matrix.
+  Step 3. solve the matrix equation self.to_2D() @ result = other.to_1D()
+  Step 4. reshape the result.
+  */
+  int64_t ndim = self.dim();
+  Tensor self_ = self;
+
+  // move dimensions of `self_` from `dims` to the end
+  if (dims.has_value()) {
+    DimVector dest_axes(dims.value().size());
+    std::iota(dest_axes.begin(), dest_axes.end(), ndim - dest_axes.size());
+    self_ = at::movedim(self_, dims.value(), dest_axes);
+  }
+
+  // result_shape is self_.sizes[-(an-other.dim):]
+  std::vector<int64_t> result_shape = self_.sizes().slice(other.dim(), ndim - other.dim()).vec();
+
+  int64_t result_product = std::accumulate(result_shape.begin(), result_shape.end(), int64_t{1}, std::multiplies<int64_t>());
+  int64_t other_product = std::accumulate(other.sizes().begin(), other.sizes().end(), int64_t{1}, std::multiplies<int64_t>());
+
+  // Check whether the self tensor can be reshaped to the 2D square matrix
+  TORCH_CHECK(result_product == other_product,
+    "Expected self to be a 'square' tensor such that prod(result.shape) == prod(other.shape), but got ",
+    result_product, " != ", other_product);
+
+  self_ = self_.reshape({result_product, result_product});
+
+  // 0th output of at::solve is the solution
+  // normally `other` would be flattened by at::solve expects 2D input
+  Tensor result = std::get<0>(at::solve(other.reshape({other.numel(), 1}), self_));
+  return result.reshape(result_shape);
+}
+
+Tensor& linalg_tensorsolve_out(Tensor& result, const Tensor& self, const Tensor& other, optional<IntArrayRef> dims) {
+  CheckedFrom c = "linalg_tensorsolve_out";
+  TensorArg result_arg(result, "result", 0);
+  TensorArg self_arg(result, "self", 1);
+  checkSameType(c, result_arg, self_arg);
+
+  Tensor self_ = self;
+  // move dimensions of `self_` from `dims` to the end
+  if (dims.has_value()) {
+    DimVector dest_axes(dims.value().size());
+    std::iota(dest_axes.begin(), dest_axes.end(), self.dim() - dest_axes.size());
+    self_ = at::movedim(self_, dims.value(), dest_axes);
+  }
+
+  auto expected_result_shape = self_.sizes().slice(other.dim(), self.dim() - other.dim());
+  TORCH_CHECK(result.sizes().equals(expected_result_shape),
+    "Expected result tensor to have size of ", expected_result_shape, ", but got tensor of size ", result.sizes());
+
+  // We've already reordered self with dims, let's pass that result
+  Tensor result_tmp = at::linalg_tensorsolve(self_, other, c10::nullopt);
+  result.copy_(result_tmp);
```

Comment:
An interesting side affect of this approach is that it will allow copying to a result tensor with a different dtype than result_tmp. This pattern of computing a result and then copying to an out tensor is common enough we should consider making a helper function for it.

Reply:
I don't know what's the expected behavior in general, so there is a check now that doesn't allow different dtypes.
https://github.com/pytorch/pytorch/pull/46142/files/497bc2241ea7ceae548277aac989849ad5a1bb0b#diff-db2b91c51cffd1fba933933ac20de256f5f0ced095890a1a49a17d73037001a2R1648